### ISSUE TITLE
[Phase 6] Semi-Supervised Learning

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/multi-label-doc-classifier#7]
## Summary

Implement semi-supervised learning to leverage the large corpus of unlabeled documents, using self-training and pseudo-labeling techniques integrated with the active learning loop.

**Parent Issue**: #1
**Depends On**: Phase 2, Phase 4

## Objectives

- Implement self-training with confidence thresholds
- Create pseudo-labeling strategy with quality monitoring
- Integrate semi-supervised learning with active learning
- Detect and prevent model degradation from bad pseudo-labels

## System Context

```
src/
└── models/
    ├── semi_supervised.py   # NEW - Self-training wrapper
    └── pseudo_labeling.py   # NEW - Pseudo-label selection

src/
└── active_learning/
    └── hybrid.py            # UPDATE - Add semi-supervised integration
```

### Self-Training
- [ ] Create `src/models/semi_supervised.py`
- [ ] Implement `SelfTrainingClassifier` wrapper
- [ ] Confidence threshold for pseudo-label acceptance
- [ ] Iterative retraining with pseudo-labels
- [ ] Track pseudo-label statistics

### Pseudo-Labeling
- [ ] Create `src/models/pseudo_labeling.py`
- [ ] Implement `PseudoLabelSelector` class
- [ ] Multiple confidence criteria (per-sample, per-label)
- [ ] Label smoothing for noisy pseudo-labels
- [ ] Curriculum learning (easy samples first)

### Quality Monitoring
- [ ] Validation set performance tracking
- [ ] Early stopping on performance degradation
- [ ] Pseudo-label confidence distribution monitoring
- [ ] Detection of confirmation bias

### Active + Semi-Supervised Integration
- [ ] Combine active learning with self-training
- [ ] Use active learning for uncertain samples
- [ ] Use pseudo-labels for confident samples
- [ ] Balance labeled vs pseudo-labeled importance

### SelfTrainingClassifier

```python
# src/models/semi_supervised.py
import numpy as np
from sklearn.base import clone
from typing import Optional, Tuple, List

class SelfTrainingClassifier:
    """Self-training wrapper for semi-supervised multi-label classification.

    Iteratively adds confident pseudo-labeled samples to the training set.
    """

    def __init__(
        self,
        base_classifier,
        confidence_threshold: float = 0.9,
        max_iterations: int = 10,
        samples_per_iteration: Optional[int] = None,
        validation_fraction: float = 0.1,
        early_stopping_patience: int = 2
    ):
        """Initialize self-training classifier.

        Args:
            base_classifier: Sklearn-compatible classifier
            confidence_threshold: Minimum confidence for pseudo-label acceptance
            max_iterations: Maximum self-training iterations
            samples_per_iteration: Max samples to add per iteration (None = all confident)
            validation_fraction: Fraction of labeled data for validation
            early_stopping_patience: Stop if no improvement for this many iterations
        """
        self.base_classifier = base_classifier
        self.confidence_threshold = confidence_threshold
        self.max_iterations = max_iterations
        self.samples_per_iteration = samples_per_iteration
        self.validation_fraction = validation_fraction
        self.early_stopping_patience = early_stopping_patience

        self.classifier_ = None
        self.training_history_ = []

    def fit(
        self,
        X_labeled: np.ndarray,
        y_labeled: np.ndarray,
        X_unlabeled: np.ndarray
    ):
        """Fit with labeled data and pseudo-label unlabeled data.

        Args:
            X_labeled: Labeled sample features
            y_labeled: Labels for labeled samples
            X_unlabeled: Unlabeled sample features
        """
        # Split validation set from labeled data
        n_val = int(len(X_labeled) * self.validation_fraction)
        if n_val > 0:
            X_val, y_val = X_labeled[:n_val], y_labeled[:n_val]
            X_train, y_train = X_labeled[n_val:], y_labeled[n_val:]
        else:
            X_val, y_val = None, None
            X_train, y_train = X_labeled, y_labeled

        # Track which unlabeled samples are still available
        unlabeled_mask = np.ones(len(X_unlabeled), dtype=bool)

        # Initial training
        self.classifier_ = clone(self.base_classifier)
        self.classifier_.fit(X_train, y_train)

        best_score = self._evaluate(X_val, y_val) if X_val is not None else 0
        patience_counter = 0

        for iteration in range(self.max_iterations):
            # Get predictions on remaining unlabeled samples
            X_pool = X_unlabeled[unlabeled_mask]
            if len(X_pool) == 0:
                break

            # Get confident pseudo-labels
            selected_local, pseudo_labels, confidences = self._select_confident(X_pool)

            if len(selected_local) == 0:
                break

            # Map local indices to global
            global_indices = np.where(unlabeled_mask)[0][selected_local]

            # Add pseudo-labeled samples to training set
            X_train = np.vstack([X_train, X_unlabeled[global_indices]])
            y_train = np.vstack([y_train, pseudo_labels])

            # Mark as used
            unlabeled_mask[global_indices] = False

            # Retrain
            self.classifier_ = clone(self.base_classifier)
            self.classifier_.fit(X_train, y_train)

            # Track progress
            current_score = self._evaluate(X_val, y_val) if X_val is not None else 0
            self.training_history_.append({
                'iteration': iteration,
                'samples_added': len(selected_local),
                'mean_confidence': np.mean(confidences),
                'validation_score': current_score,
                'total_training_samples': len(X_train)
            })

            # Early stopping
            if current_score > best_score:
                best_score = current_score
                patience_counter = 0
            else:
                patience_counter += 1
                if patience_counter >= self.early_stopping_patience:
                    print(f"Early stopping at iteration {iteration}")
                    break

        return self

    def _select_confident(
        self,
        X_pool: np.ndarray
    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
        """Select samples with confident predictions.

        Returns:
            Tuple of (indices, pseudo_labels, confidences)
        """
        if hasattr(self.classifier_, 'predict_proba'):
            probas = self.classifier_.predict_proba(X_pool)
            if isinstance(probas, list):
                probas = np.column_stack([p[:, 1] if p.ndim > 1 else p for p in probas])
        else:
            probas = self.classifier_.predict(X_pool).astype(float)

        # Confidence = distance from 0.5
        confidences_per_label = np.abs(probas - 0.5) * 2  # Scale to [0, 1]
        min_confidence = np.min(confidences_per_label, axis=1)

        # Select confident samples
        confident_mask = min_confidence >= self.confidence_threshold
        confident_indices = np.where(confident_mask)[0]

        if self.samples_per_iteration is not None:
            # Limit number of samples
            if len(confident_indices) > self.samples_per_iteration:
                # Take most confident
                top_indices = np.argsort(min_confidence[confident_mask])[-self.samples_per_iteration:]
                confident_indices = confident_indices[top_indices]

        # Generate pseudo-labels (threshold at 0.5)
        pseudo_labels = (probas[confident_indices] >= 0.5).astype(int)
        sample_confidences = min_confidence[confident_indices]

        return confident_indices, pseudo_labels, sample_confidences

    def _evaluate(self, X_val: np.ndarray, y_val: np.ndarray) -> float:
        """Evaluate on validation set."""
        if X_val is None:
            return 0
        predictions = self.classifier_.predict(X_val)
        # Subset accuracy
        return np.mean(np.all(predictions == y_val, axis=1))

    def predict(self, X) -> np.ndarray:
        """Predict labels."""
        return self.classifier_.predict(X)

    def predict_proba(self, X):
        """Predict probabilities."""
        return self.classifier_.predict_proba(X)
```

### PseudoLabelSelector

```python
# src/models/pseudo_labeling.py
import numpy as np
from typing import Tuple, Optional

class PseudoLabelSelector:
    """Selects high-quality pseudo-labels with various strategies."""

    def __init__(
        self,
        strategy: str = 'threshold',
        confidence_threshold: float = 0.9,
        top_k: Optional[int] = None,
        curriculum: bool = False
    ):
        """Initialize selector.

        Args:
            strategy: 'threshold', 'top_k', or 'curriculum'
            confidence_threshold: For threshold strategy
            top_k: For top_k strategy
            curriculum: Start with easy samples, increase difficulty
        """
        self.strategy = strategy
        self.confidence_threshold = confidence_threshold
        self.top_k = top_k
        self.curriculum = curriculum
        self._iteration = 0

    def select(
        self,
        probas: np.ndarray,
        previously_selected: Optional[np.ndarray] = None
    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
        """Select samples for pseudo-labeling.

        Args:
            probas: Prediction probabilities (n_samples, n_labels)
            previously_selected: Mask of already selected samples

        Returns:
            (indices, pseudo_labels, confidences)
        """
        # Compute sample-level confidence
        confidence_per_label = np.maximum(probas, 1 - probas)
        sample_confidence = np.min(confidence_per_label, axis=1)

        # Exclude previously selected
        if previously_selected is not None:
            sample_confidence[previously_selected] = 0

        if self.curriculum:
            # Increase threshold over time (start easy)
            effective_threshold = self.confidence_threshold - 0.1 * max(0, 5 - self._iteration)
            effective_threshold = max(0.5, effective_threshold)
        else:
            effective_threshold = self.confidence_threshold

        if self.strategy == 'threshold':
            selected = sample_confidence >= effective_threshold

        elif self.strategy == 'top_k':
            k = self.top_k or 100
            if np.sum(sample_confidence > 0) < k:
                k = np.sum(sample_confidence > 0)
            if k == 0:
                return np.array([]), np.array([]).reshape(0, probas.shape[1]), np.array([])
            threshold = np.partition(sample_confidence, -k)[-k]
            selected = sample_confidence >= threshold

        else:
            raise ValueError(f"Unknown strategy: {self.strategy}")

        indices = np.where(selected)[0]
        pseudo_labels = (probas[indices] >= 0.5).astype(int)
        confidences = sample_confidence[indices]

        self._iteration += 1

        return indices, pseudo_labels, confidences

    def reset(self):
        """Reset iteration counter for curriculum learning."""
        self._iteration = 0
```

## Verification

```python
from src.models.semi_supervised import SelfTrainingClassifier
from sklearn.linear_model import LogisticRegression

# Setup
base_clf = LogisticRegression(max_iter=1000)
semi_clf = SelfTrainingClassifier(
    base_clf,
    confidence_threshold=0.85,
    max_iterations=5
)

# Train with labeled + unlabeled
semi_clf.fit(X_labeled, y_labeled, X_unlabeled)

# Check history
for record in semi_clf.training_history_:
    print(f"Iteration {record['iteration']}: "
          f"added {record['samples_added']} samples, "
          f"validation: {record['validation_score']:.3f}")

# Verify improvement over supervised-only
supervised_clf = LogisticRegression(max_iter=1000)
supervised_clf.fit(X_labeled, y_labeled)

semi_accuracy = np.mean(np.all(semi_clf.predict(X_test) == y_test, axis=1))
supervised_accuracy = np.mean(np.all(supervised_clf.predict(X_test) == y_test, axis=1))

print(f"Supervised: {supervised_accuracy:.3f}")
print(f"Semi-supervised: {semi_accuracy:.3f}")
assert semi_accuracy >= supervised_accuracy - 0.02, "Semi-supervised should not be much worse"
```

## Definition of Done

- [ ] Self-training improves or maintains performance vs supervised-only
- [ ] Confidence threshold effectively filters low-quality pseudo-labels
- [ ] Early stopping prevents degradation from bad pseudo-labels
- [ ] Training history provides visibility into the process
- [ ] Integrates cleanly with active learning loop
- [ ] Unit tests achieve 80% coverage

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/models/semi_supervised.py` | - | Create | Self-training classifier |
| `src/models/pseudo_labeling.py` | - | Create | Pseudo-label selector |
| `src/active_learning/hybrid.py` | - | Update | Add semi-supervised |
| `tests/test_semi_supervised.py` | - | Create | Self-training tests |
| `tests/test_pseudo_labeling.py` | - | Create | Selector tests |
